[Event Hubs] Update RetryOptions interface to pass exponential retry params

ramya0820 opened the pull request:
For more context refer to #4266

chradek commented on sdk/eventhub/event-hubs/src/eventHubReceiver.ts:
```diff
@@ -531,7 +531,7 @@ export class EventHubReceiver extends LinkEntity {
         operation: () => this.initialize(initOptions),
         operationType: RetryOperationType.receiverLink,
-        maxRetries: Constants.defaultMaxRetriesForConnection
+        maxRetries: Constants.defaultMaxRetriesForConnection,
```
Super nit: looks like this change wasn't needed, does running `npm run format` remove the comma?

ramya0820 replied:
That's odd, it wasn't going away on file save. After merge from master it disappears on file save itself..

chradek commented on sdk/eventhub/event-hubs/src/receiver.ts:
```diff
@@ -448,9 +452,18 @@ export class EventHubConsumer {
 
+    const timeoutInMs = retryOptions.timeoutInMs;
+    const retryPolicy = retryOptions.retryPolicy;
+    const minExponentialRetryDelayInMs = retryOptions.minExponentialRetryDelayInMs;
```
Are there any default values for the new exponential retry fields you're adding? Will `retry` throw any errors if they are missing but we've specified an `exponential` retry policy?

ramya0820 replied:
Yeah, default values are being set https://github.com/Azure/azure-sdk-for-js/blob/master/sdk/core/core-amqp/src/retry.ts#L160-L171
So nope, no errors will be thrown if just the the `retryPolicy` is selected to be as exponential

ramya-rao-a commented on sdk/eventhub/event-hubs/src/receiver.ts:
```diff
@@ -438,7 +442,7 @@ export class EventHubConsumer {
   private _initRetryOptions(
     retryOptions: RetryOptions = {}
-  ): Required<Pick<RetryOptions, "maxRetries" | "retryInterval">> {
+  ): Required<Pick<RetryOptions, "maxRetries" | "retryInterval">> & RetryOptions {
```
@chradek What was the intention of `_initRetryOptions`?
I initially thought it was a way to guarantee non undefined values on `maxRetries` and `retryInterval`

With this change, we now allow the rest of properties of RetryOptions here that may be undefined.
I have this feeling that the purpose is being defeated if `initRetryOptions` is not initializing with some value

chradek replied:
@ramya-rao-a 
The intention was indeed to make sure `maxRetries` and `retryInterval` were always defined since we have known default options for them. I think the method name is fine...a comment could be added to indicate that it will set default values if a default exists.

With that in mind, I think this should also set `retryPolicy` to required since we have a known default (Linear)

ramya0820 replied:
Sure, we can add the comment for clarity

About default for `retryPolicy`, `core-amqp` handles it at https://github.com/Azure/azure-sdk-for-js/blob/master/sdk/core/core-amqp/src/retry.ts#L217

But agree that the value can be set and made required - not one way or against other since this is an internal utility. But we might want to rethink about why we need this utility and have any values be marked as required when they all by design seem to be optional.

chradek replied:
This method was originally suggested so that we wouldn't have to keep calculating defaults every time a receiveBatch operation was called.

Since we are setting the defaults once when instantiating the class, we also know that these fields are always set. Making them appear `Required` lets us skip the null checks and avoid using the non-null assertion.

ramya0820 replied:
Got it, that makes sense in this context
Although I think the defaults for these retry related options can be managed, set within `core-amqp`'s retry?
It doesn't seem like there's much value in doing it again, unless these were Event Hubs specific defaults~

ramya0820 replied:
About the non-null assertion argument, wondering if that can be then solved differently, by making all internal usage of options mandatory perhaps? We can track this refactoring in a different issue - just that the whole manipulation of what's "Required<> && optional"  etc. seems like it can be simplified preferably

ramya-rao-a replied:
`core-amqp` does set default values
In Event Hubs, we don't need to do null checks on the properties of the `retryOptions` and can confidently pass `retryOptions.maxRetries`, `retryOptions.retryInterval` etc to the `retry()` function of `core-amqp`.

The null check on `retryOptions` itself can be avoided by initializing it to `{}` if no retry options are provided. This way we can use `retryOptions.maxRetries` directly instead of `retryOptions && retryOptions.maxRetries`.

So, I am all for having the retry options as a private variable in classes like `EventHubConsumer`, `EventHubProducer`, `EventHubSender` etc and have it be non nullable. But we don't need to ensure that the internal properties of retry options are non nullable as well.

ramya0820 replied:
I've updated  `receiver` - to simplify and remove the `initRetryOptions`

Based on this thread, it looks like the utilities we introduced for SDK side validation such as `getRetryAttemptTimeoutInMs` for `retryOptions.timeOutinMs` can also be removed and just allow for these values to be set and passed in by the user? 
Because the default values are not SDK specific and are already being computed and set within retry? @ramya-rao-a

ramya-rao-a replied:
Thanks, `receiver.ts` looks good

ramya0820 replied:
- About classes like `EventHubSender`, since we are passing in the `retryOptions` as part of `senderOptions` current design/implementation doesn't seem to benefit from using a private instance. Because, we anyway we need to check contents of `senderOptions` from within `send` implementations that setup, perform retries.
- If we do want to have one, then we'll need to refactor the internal API to not require passing `senderOptions` / `retryOptions` to `eventHubSender` and use the private instance set in `receiver` by moving implementation of `send` request and retries to `sender` as opposed to `eventHubSender` (Because for `receive`, we have it in `receiver` and not in `eventHubReceiver`.) We could refactor, improve the implementation as part of a different issue / PR.

Updated implementations for these however, to simplify and move out the checks where applicable.
Some instances of `timeOutInMs` still require default value computation as they are used in `setTimeout`. Few others such `retryInterval` need checks since they are in a differing units (seconds vs milliseconds) and can be updated as part of #4277 

(Could consider naming it to be same as what we have for retry in `core-amqp`  to likes of `delayInMs`?) @ramya-rao-a @bterlson

ramya0820 replied:
Discussed offline with @bterlson, we would like to update `retryInterval` to be `delayInSeconds`
@ramya-rao-a let us know if there's any concern with it.

ramya-rao-a replied:
@ramya0820 I would prefer not to do the renames of the retry options in this PR. Let's scope this PR to
- passing the exponential retry related options from user to relevant places
- ensuring that this flow from user to the relevant places are using good coding practices

ramya-rao-a replied:
> Few others such retryInterval need checks since they are in a differing units (seconds vs milliseconds) and can be updated as part of #4277

Agreed that we need to do the conversion and so need checks, but we dont need to set the defaults as we expect `core-amqp` to do that. So I would suggest something like the below

`typeof retryInterval === "number" ? retryInterval/1000 : undefined`

chradek commented on sdk/eventhub/event-hubs/src/receiver.ts:
```diff
@@ -448,9 +452,18 @@ export class EventHubConsumer {
         : Constants.defaultDelayBetweenOperationRetriesInSeconds;
 
+    const timeoutInMs = retryOptions.timeoutInMs;
+    const retryPolicy = retryOptions.retryPolicy;
```
We stated that `retryPolicy` will default to `LinearRetryPolicy` in the docs. This would be a good place to set it if it is currently undefined.

chradek commented on sdk/eventhub/event-hubs/src/receiver.ts:
```diff
@@ -73,7 +73,8 @@ export class EventHubConsumer {
    * @property The set of retry options to configure the receiveBatch operation.
    */
-  private _retryOptions: Required<Pick<RetryOptions, "maxRetries" | "retryInterval">>;
+  private _retryOptions: Required<Pick<RetryOptions, "maxRetries" | "retryInterval">> &
```
Since we can set a default value for `retryPolicy`, it should also be a required field.

chradek commented on sdk/eventhub/event-hubs/src/receiver.ts:
```diff
@@ -370,7 +371,7 @@ export class EventHubConsumer {
 
         const addTimeout = (): void => {
-          let msg = "[%s] Setting the wait timer for %d seconds for receiver '%s'.";
+          const msg = "[%s] Setting the wait timer for %d seconds for receiver '%s'.";
```
👍

ramya0820 commented:
> * In the `getMaxMessageSize()` is a function that uses retry. It should also use the exponential parameters as well

Updated

> * I like what we have done in the `receiver.ts` file i.e have all the defaults handled in core-amqp and avoid the `retryOptions && retryOptions.maxRetries` model in favor of `retryOptions.amxRetries`. Let's ensure we do the same in the `eventHubSender.ts` file

Updated

> * I believe #4322 is almost done. Let's get that merged, and then update this PR to pass in the exponential things to the management operations

Updated, yes depending on which PR gets merged first, I can take care of the conflicts and ensure this is in place. (The other one's waiting on another approval)


cc: @ramya-rao-a

ramya-rao-a commented on sdk/eventhub/event-hubs/src/receiver.ts:
```diff
@@ -404,10 +404,13 @@ export class EventHubConsumer {
       connectionHost: this._context.config.host,
       connectionId: this._context.connectionId,
-      delayInSeconds: retryOptions.retryInterval,
+      delayInSeconds: retryOptions.delayInMs,
```
We are passing milliseconds to something that expects seconds

ramya-rao-a commented on sdk/eventhub/event-hubs/src/managementClient.ts:
```diff
@@ -325,9 +325,9 @@ export class ManagementClient extends LinkEntity {
-            ? options.retryOptions.retryInterval / 1000
+          options.retryOptions.delayInMs &&
+          options.retryOptions.delayInMs >= 0
+            ? options.retryOptions.delayInMs / 1000
```
In the spirit of avoiding the null check for `retryOptions` we should initialize `options.retryOptions` to `{}` if no value was provided like we are doing in a couple of other places in this PR

ramya0820 replied:
Addressed this and other comments, please see latest commit

ramya-rao-a commented on sdk/eventhub/event-hubs/src/eventHubSender.ts:
```diff
@@ -742,19 +743,20 @@ export class EventHubSender extends LinkEntity {
-        : Constants.defaultDelayBetweenOperationRetriesInSeconds;
+      typeof retryOptions.retryInterval === "number"
+        ? retryOptions.retryInterval / 1000
+        : undefined;
```
The above 2 constants are not needed. They can be inlined just like we do in other places where we create the `RetryConfig`

ramya-rao-a replied:
Then as part of #4401, the inline will get simplified too

ramya-rao-a commented on sdk/eventhub/event-hubs/src/managementClient.ts:
```diff
@@ -414,13 +415,11 @@ export class ManagementClient extends LinkEntity {
-          : Constants.defaultDelayBetweenOperationRetriesInSeconds;
+        typeof retryOptions.retryInterval === "number"
+          ? retryOptions.retryInterval / 1000
+          : undefined;
```
The above 2 constants are not needed. They can be inlined just like we do in other places where we create the `RetryConfig`. Then as part of #4401, the inline will get simplified too

ramya0820 commented:
> Looks good. Just 2 comments regarding inlining `maxRetries` and `delayInSeconds` to be consistent

Updated
